source/_posts/2019-01-28-building-a-decentralized-reddit-with-embark-part-1.md

```diff
@@ -0,0 +1,192 @@
+
+<pre class="solidity"><code>struct Post {
+  ...
+  uint upvotes;
+  uint downvotes;
+  mapping(address => Ballot) voters;
+}</code></pre>
+
+Here's the `NewPost` event which we'll use in a few moments:
+
+<pre class="solidity"><code>event NewVote(
+  uint indexed postId,
+  address owner,
+  uint8 vote
+);</code></pre>
+
+Last but not least, we have to update our `createPost()` function as the `Post` struct now needs `upvotes` and `downvotes`:
+
+
+<pre class="solidity"><code>function createPost(bytes _description) public {
+  ...
+  posts[postId] = Post({
+    ...
+    upvotes: 0,
+    downvotes: 0
+  });
+}</code></pre>
+
+With these building blocks at hand, let's implement a `vote(uint postId, uint8 _vote)` method. `_vote` is going to be one of our defined `Ballot` types and is represented as uint going from 0 - 2. We'll use Solidity's `require()` statement to ensure we only vote on posts that actually exist, as well as nobody can actually vote multiple times on the same post. 
+
+We then increment the up or down vote counter respectively, store the voter and emit a `NewVote` event:
+
+<pre class="solidity"><code>function vote(uint _postId, uint8 _vote) public {
+  Post storage post = posts[_postId];
+
+  require(post.creationDate != 0, "Post does not exist");
+  require(post.voters[msg.sender] == Ballot.NONE, "You already voted on this post");
+
+  Ballot ballot = Ballot(_vote);
+
+  if (ballot == Ballot.UPVOTE) {
+      post.upvotes++;
+  } else {
+      post.downvotes++;
+  }
+
+  post.voters[msg.sender] = ballot;
+  emit NewVote(_postId, msg.sender, _vote);
+}</code></pre>
+
+### Determine if users can vote
+
+We probably want to add an indication to the UI that a user has already voted on a certain post. For that it'd be handy to have an API that actually tells us whether a user can vote on a post. We've already discussed earlier that users can't vote multiple times on the same post, so figuring out if a user can vote is pretty straight forward. Here's what a `canVote(uint _postId)` method could look like:
+
+<pre class="solidity"><code>function canVote(uint _postId) public view returns (bool) {
+  if (_postId > posts.length - 1) return false;
+  Post storage post = posts[_postId];
+  return (post.voters[msg.sender] == Ballot.NONE);
+}</code></pre>
+
+### Fetching votes
+
+We also need a way to actually let users check what they've voted for, in case they did. For that we'll add a simple `getVote()` method that looks something like this:
+
+<pre class="solidity"><code>function getVote(uint _postId) public view returns (uint8) {
+  Post storage post = posts[_postId];
+  return uint8(post.voters[msg.sender]);
+}</code></pre>
+
+And with that, our Smart Contract is pretty much done! Just to make sure that everything is compiling smoothly, we can execute `embark run` in case there's no existing instance watching our work already.
```

Comment:
Since this is only smart contract, you can recommend using `embark build --contract`